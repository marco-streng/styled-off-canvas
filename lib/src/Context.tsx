import * as React from "react";

type Action = { type: "toggle" | "close" };
type Dispatch = (action: Action) => void;
type State = { isOpen: boolean };
type OffCanvasProviderProps = { children: React.ReactNode };

const OffCanvasStateContext = React.createContext<
  { state: State; dispatch: Dispatch } | undefined
>(undefined);

const offCanvasReducer = (state: State, action: Action) => {
  switch (action.type) {
    case "toggle": {
      return { isOpen: !state.isOpen };
    }

    case "close": {
      return { isOpen: false };
    }

    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

const OffCanvasProvider = ({ children }: OffCanvasProviderProps) => {
  const [state, dispatch] = React.useReducer(offCanvasReducer, {
    isOpen: false,
  });

  const value = { state, dispatch };

  return (
    <OffCanvasStateContext.Provider value={value}>
      {children}
    </OffCanvasStateContext.Provider>
  );
};

const useOffCanvas = () => {
  const context = React.useContext(OffCanvasStateContext);
  if (context === undefined) {
    throw new Error("useOffCanvas must be used within a OffCanvasProvider");
  }
  return context;
};

export { OffCanvasProvider, useOffCanvas };
