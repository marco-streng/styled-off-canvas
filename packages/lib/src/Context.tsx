import * as React from "react";

type Action = { type: "toggle" | "close" | "open" };

type State = { isOpen: boolean; onClose?: () => void };

type OffCanvasProviderProps = {
  children: React.ReactNode;
  isOpen?: boolean;
  onClose?: () => void;
};

const OffCanvasStateContext = React.createContext<
  | {
      isOpen: boolean;
      onClose?: () => void;
      close: () => void;
      open: () => void;
      toggle: () => void;
    }
  | undefined
>(undefined);

const offCanvasReducer = (state: State, action: Action) => {
  switch (action.type) {
    case "toggle": {
      if (state.isOpen && state.onClose) {
        state.onClose();
      }

      return { ...state, isOpen: !state.isOpen };
    }

    case "open": {
      return { ...state, isOpen: true };
    }

    case "close": {
      if (state.onClose) {
        state.onClose();
      }

      return { ...state, isOpen: false };
    }

    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

const OffCanvasProvider = ({
  children,
  isOpen,
  onClose,
}: OffCanvasProviderProps) => {
  const [state, dispatch] = React.useReducer(offCanvasReducer, {
    isOpen: isOpen ?? false,
    onClose,
  });

  const value = {
    isOpen: state.isOpen,
    onClose: state.onClose,
    close: () => dispatch({ type: "close" }),
    open: () => dispatch({ type: "open" }),
    toggle: () => dispatch({ type: "toggle" }),
  };

  return (
    <OffCanvasStateContext.Provider value={value}>
      {children}
    </OffCanvasStateContext.Provider>
  );
};

const useOffCanvas = () => {
  const context = React.useContext(OffCanvasStateContext);
  if (context === undefined) {
    throw new Error("useOffCanvas must be used within an OffCanvasProvider");
  }
  return context;
};

export { OffCanvasProvider, useOffCanvas };
