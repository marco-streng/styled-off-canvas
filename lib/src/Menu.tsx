import React from "react";
import styled from "styled-components";
import { useOffCanvas } from "./Context";

const Container = styled.div<{
  $isOpen: boolean;
  $background?: string;
  $duration?: string;
  $position?: "left" | "right";
  $width?: string;
}>(
  ({
    $isOpen,
    $background = "#fff",
    $duration = "500ms",
    $position = "right",
    $width = "300px",
  }) => `
  background-color: ${$background};
  height: 100%;
  left: ${$position === "left" ? ($isOpen ? 0 : "-100%") : "inherit"};
  max-width: ${$width};
  overflow-y: auto;
  position: fixed;
  right: ${$position === "right" ? ($isOpen ? 0 : "-100%") : "inherit"};
  top: 0;
  transition: ${$position} ${$duration} ease-in-out;
  width: 100%;
  z-index: 10;
`
);

const isCSR = typeof window !== "undefined";

const initalBodyStyle = {
  maxHeight: isCSR ? window.document.body.style.maxHeight : "none",
  overflow: isCSR ? window.document.body.style.overflow : "visbile",
};

/**
 * Menu component
 *
 * @param   {string}    className styled-components className for custom styling
 * @returns {component}           React component
 */
export const Menu = ({
  children,
  background,
  duration,
  position,
  width,
  closeOnEsc,
  className,
  ...props
}: {
  children: React.ReactNode;
  background?: string;
  duration?: string;
  position?: "left" | "right";
  width?: string;
  closeOnEsc?: boolean;
  className?: string;
}) => {
  const { isOpen, close } = useOffCanvas();

  // Avoid scrolling on content when the navigation is open
  React.useEffect(() => {
    const bodyElStyle = window.document.body.style;

    if (isOpen) {
      bodyElStyle.maxHeight = "100%";
      bodyElStyle.overflow = "hidden";
    } else {
      bodyElStyle.maxHeight = initalBodyStyle.maxHeight;
      bodyElStyle.overflow = initalBodyStyle.overflow;
    }
  }, [isOpen]);

  if (closeOnEsc) {
    React.useEffect(() => {
      const handleKeyDown = (event: KeyboardEvent) => {
        if (event.code.toLowerCase() === "escape") {
          close();
        }
      };

      isOpen
        ? window.document.addEventListener("keydown", handleKeyDown)
        : window.document.removeEventListener("keydown", handleKeyDown);
    }, [isOpen]);
  }

  return (
    <Container
      className={className}
      $background={background}
      $duration={duration}
      $isOpen={isOpen}
      $position={position}
      $width={width}
      {...props}
    >
      {children}
    </Container>
  );
};
