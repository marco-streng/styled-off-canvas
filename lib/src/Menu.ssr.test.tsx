// @vitest-environment node

import React from "react";
import TestRenderer from "react-test-renderer";
import { SpyInstance } from "vitest";
import { OffCanvasProvider } from "./Context";
import { Menu } from "./Menu";

describe("Menu", () => {
  let useEffectSpy: SpyInstance;

  beforeEach(() => {
    // useEffect won't run on server
    useEffectSpy = vi.spyOn(React, "useEffect");
    useEffectSpy.mockImplementation(() => vi.fn());
  });

  afterEach(() => {
    useEffectSpy.mockRestore();
  });

  test("ssr", () => {
    const tree = TestRenderer.create(
      <OffCanvasProvider isOpen>
        <Menu>
          <div>content</div>
        </Menu>
      </OffCanvasProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
