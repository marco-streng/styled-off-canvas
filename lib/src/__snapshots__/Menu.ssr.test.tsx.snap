// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > ssr 1`] = `
<div
  className="sc-bcPKhP geetHI"
>
  <div>
    content
  </div>
</div>
`;
